### PR TITLE
Implement socket.io for chat

### PR DESCRIPTION
**Pull request by huyanh7796**

### CODE REVIEW

File: services/socket.js [hunk 1]
```diff
@@ -25,7 +26,35 @@ const activateSocket = (io) => {
...
+  io.use(async function (socket, next) {
+    try {
+      if (socket.handshake.query.token) {
+        const token = socket.handshake.query.token;
+        const data = await jwt.verify(token, config.SECRET_KEY);
+  
+        if (!data) {
+          return next(new Error("Not authenticated!"));
+        }
+
```
Comment (nddmanh) [inline services/socket.js]: I think you should delete this line, because it looks sporadic.

File: services/socket.js [hunk 2]
```diff
@@ -25,7 +26,35 @@ const activateSocket = (io) => {
...
+        const data = await jwt.verify(token, config.SECRET_KEY);
+  
+        if (!data) {
+          return next(new Error("Not authenticated!"));
+        }
+  
+        if (data.exp <= Date.now() / 1000) {
+          return next(new Error("Token expired!"));
+        }
+
```
Comment (nddmanh) [inline services/socket.js]: here too

File: services/socket.js [hunk 3]
```diff
@@ -25,7 +26,35 @@ const activateSocket = (io) => {
...
+        if (data.exp <= Date.now() / 1000) {
+          return next(new Error("Token expired!"));
+        }
+  
+        socket.user = {
+          _id: data._id,
+          username: data.username,
+          role: data.role
+        };
+
```
Comment (nddmanh) [inline services/socket.js]: here too

File: services/socket.js [hunk 4]
```diff
@@ -25,7 +26,35 @@ const activateSocket = (io) => {
...
+  
+        socket.user = {
+          _id: data._id,
+          username: data.username,
+          role: data.role
+        };
+  
+        return next();
+      }
+
```
Comment (nddmanh) [inline services/socket.js]: here too

File: services/socket.js [hunk 5]
```diff
@@ -38,43 +40,32 @@ const activateSocket = (io) => {
...
-        return next();
-      }
-      return next(new Error("Not authenticated!"));
-    } catch (err) {
-      return next(err);
-    }
-  }).on('connection', (socket) => {
+  io.use((socket, next) => socketAuth(socket, next)).on('connection', (socket) => {
+    io.emit('connect-success', {
+      _id : socket.user._id,
```
Comment (lethemanh) [inline services/socket.js]: Should you use another key name for more meaning instead of ```_id```

File: services/socket.js [hunk 6]
```diff
@@ -38,43 +40,32 @@ const activateSocket = (io) => {
...
       timeInterval = setInterval(countdown, COUNTDOWN_INTERVAL_SECOND);
-    }
+    };
+    socket.on('message', (message) => {
+      io.emit('chat-message', message);
+    });
     socket.on('disconnect', () => {
+      io.emit('disconnect-success', 
+      {
+        _id : socket.user._id,
```
Comment (lethemanh) [inline services/socket.js]: Here too

File: services/socket.js [hunk 7]
```diff
@@ -1,9 +1,11 @@
 const rollResult = require('../helper/rollResult');
 const calculateMultipleResults = require('../helper/calculateMultipleResults');
-const { SECRET_KEY, TIME_LEFT_DEFAULT, PENDING_COUNTDOWN_TIME, COUNTDOWN_INTERVAL_SECOND } = require('../config');
-const jwt = require('jsonwebtoken');
+const { TIME_LEFT_DEFAULT, PENDING_COUNTDOWN_TIME, COUNTDOWN_INTERVAL_SECOND } = require('../config');
+const socketAuth = require('./socketAuth');
+const MSGTYPE = require('../constants/msgtype');
 const _ = require('lodash');
 
+
```
Comment (nddmanh) [inline services/socket.js]: Is this line unnecessary?

**Review by nddmanh — APPROVED** [review]

**Pull request closed without merging**